syncapi/streams/stream_invite.go

```diff
@@ -74,27 +74,43 @@ func (p *InviteStreamProvider) IncrementalSync(
 		return to
 	}
 	for roomID := range retiredInvites {
-		if _, ok := req.Response.Rooms.Invite[roomID]; ok {
+		if req.Response.Rooms.Invite[roomID] != nil {
 			continue
 		}
-		if _, ok := req.Response.Rooms.Join[roomID]; ok {
+		if req.Response.Rooms.Join[roomID] != nil {
 			continue
 		}
-		lr := types.NewLeaveResponse()
-		h := sha256.Sum256(append([]byte(roomID), []byte(strconv.FormatInt(int64(to), 10))...))
-		lr.Timeline.Events = append(lr.Timeline.Events, gomatrixserverlib.ClientEvent{
-			// fake event ID which muxes in the to position
-			EventID:        "$" + base64.RawURLEncoding.EncodeToString(h[:]),
-			OriginServerTS: gomatrixserverlib.AsTimestamp(time.Now()),
-			RoomID:         roomID,
-			Sender:         req.Device.UserID,
-			StateKey:       &req.Device.UserID,
-			Type:           "m.room.member",
-			Content:        gomatrixserverlib.RawJSON(`{"membership":"leave"}`),
-		})
-		req.Response.Rooms.Leave[roomID] = lr
 
+		joinedUsers, err := snapshot.AllJoinedUsersInRoom(ctx, []string{roomID})
```

Comment:
This is a rather expensive way of approaching this problem as it results in us pulling out the entire membership state for a room, only to discard the vast majority of it. The membership table already has `SelectMembershipForUser` so it seems like we could just add some wrapper for that into `storage_sync.go`.

Reply:
thanks, this is the kind of feedback I'm hoping to get. I can work on a better fix in storage_sync.  I'm investigating a second regression that doesn't sync space room state properly after a user accepts an invite and joins a room. All happening from the commits in the last week - all related to invite-join flow.  Missing test coverage?

Reply:
@neilalexander Fixed to use SelectMembershipForUser. Please re-review